src/collapse/vsprof.rs

```diff
@@ -0,0 +1,188 @@
+use std::{
+    cmp::Ordering,
+    io::{self, BufRead},
+};
+
+use log::warn;
+
+use crate::collapse::{common::Occurrences, Collapse};
+
+static START_LINE: &str = "Level,Function Name,Number of Calls,Elapsed Inclusive Time %,Elapsed Exclusive Time %,Avg Elapsed Inclusive Time,Avg Elapsed Exclusive Time,Module Name,";
+
+/// A stack collapser for the output of the built in Visual Studio profiler.
+#[derive(Default)]
+pub struct Folder {
+    /// Function entries on the stack in this entry thus far.
+    stack: Vec<(String, usize)>,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: std::io::BufRead,
+        W: std::io::Write,
+    {
+        // Skip the header
+        let mut line = Vec::new();
+        if reader.read_until(b'\n', &mut line)? == 0 {
+            warn!("File ended before start of call graph");
+            return Ok(());
+        };
+
+        // Process the data
+        let mut occurences = Occurrences::new(1);
+        loop {
+            line.clear();
+            if reader.read_until(b'\n', &mut line)? == 0 {
+                break;
+            }
+            let l = String::from_utf8_lossy(&line);
+            let line = l.trim_end();
+            if line.is_empty() {
+                continue;
+            } else {
+                self.on_line(line, &mut occurences)?;
+            }
+        }
+
+        // Write the results
+        occurences.write_and_clear(writer)?;
+
+        // Reset the state
+        self.stack.clear();
+        Ok(())
+    }
+
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        if let Ok(n) = input.read_line(&mut line) {
+            if n == 0 {
+                return Some(false);
+            }
+        } else {
+            return Some(false);
+        }
+
+        // Some files may start with the <U+FEFF> character (zero width no-break space). This
+        // causes the call to `starts_with` to return false, which in this case isn't what we want.
+        // As this character has no influence on the rest of the file, we can safely ignore it.
+        if line.starts_with(START_LINE) || line.starts_with(&("\u{feff}".to_owned() + START_LINE)) {
+            Some(true)
+        } else {
+            Some(false)
+        }
+    }
+}
+
+impl Folder {
+    // Parse lines like:
+    // Level,Function Name,Number of Calls,...
+    // 6,"System.String.IsNullOrEmpty(string)",4,0.00,0.00,0.00,0.00,"mscorlib.dll",
+    fn on_line(&mut self, line: &str, occurences: &mut Occurrences) -> io::Result<()> {
+        let (depth, remainder) = get_next_number(line)?;
+        // After the depth is parsed from `line`, it still has a leading comma and double
+        // quote. By trimming those, we know that the first double quote is the closing double
```

Comment:
This assumes that the function name is _always_ in double quotes. Is that the case? I could imagine that maybe it isn't if it doesn't contain spaces or something, but don't know for sure?

If it is the case, I'd use `.strip_prefix(",\"")` here over `trim_start_matches`, since the latter will remove multiple occurrences.

Reply:
I checked, the function name is always in double quotes, regardless of the presence of any spaces.